### PR TITLE
[documentation] How to debug memory leaks

### PR DESCRIPTION
**Pull request by violetagg**

### CODE REVIEW

**Review by pderop — APPROVED** [review]

**Comment by violetagg** [comment]
@Buzzardo When you have time can you review this documentation? Thanks in advance!

File: docs/asciidoc/faq.adoc [hunk 1]
```diff
@@ -85,3 +86,18 @@ Format
 Reactor Netty adds information for the connection at the beginning of every log record (when this is possible).
 Use the id of the connection in order to extract all log records for a particular HTTP request.
 For more information see <<faq.logging-prefix>>
+
+[[faq.memory-leaks]]
+== How can I debug a memory leak?
+By default, Reactor Netty uses a direct memory as this is more performant
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc:92]: Remove "a" from "a direct memory".

File: docs/asciidoc/faq.adoc [hunk 2]
```diff
@@ -85,3 +86,18 @@ Format
...
+when there are many native I/O operations (working with sockets) as this can remove the copying operations.
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc:93]: Add a comma after "(working with sockets)".

File: docs/asciidoc/faq.adoc [hunk 3]
```diff
@@ -85,3 +86,18 @@ Format
...
+As allocation and deallocation are expensive operations, the buffers, used in Reactor Netty by default, are pooled objects.
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc:94]: Change the commas around "used in Reactor Netty by default" to parentheses. While technically correct, with other commas nearby, it is hard to parse. Replacing those two commas with parentheses will clarify the sentence.

File: docs/asciidoc/faq.adoc [hunk 4]
```diff
@@ -85,3 +86,18 @@ Format
...
+For more information see https://github.com/netty/netty/wiki/Reference-counted-objects[Reference Counted Objects].
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc:95]: Add a comma after "information".

File: docs/asciidoc/faq.adoc [hunk 5]
```diff
@@ -85,3 +86,18 @@ Format
...
+
+In order to be able to debug memory issues with the direct memory and the pooled buffers, Netty provides a special memory leak detection mechanism.
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc:97]: Remove "In order". This is a general rule. It is nearly always unnecessary.

File: docs/asciidoc/faq.adoc [hunk 6]
```diff
@@ -85,3 +86,18 @@ Format
...
+Follow the instructions https://github.com/netty/netty/wiki/Reference-counted-objects#troubleshooting-buffer-leaks[Troubleshooting Buffer Leaks]
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc:98]: Add "for" between "instructions" and the link.

File: docs/asciidoc/faq.adoc [hunk 7]
```diff
@@ -85,3 +86,18 @@ Format
...
+in order to enable this mechanism. In addition to the instructions provided by Netty, Reactor Netty provides a special
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc:99]: Remove "in order".

File: docs/asciidoc/faq.adoc [hunk 8]
```diff
@@ -85,3 +86,18 @@ Format
...
+logger `_reactor.netty.channel.LeakDetection` that helps to identify where the memory leak might be located inside Reactor Netty
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc:100]: Add parentheses around `_reactor.netty.channel.LeakDetection`.

File: docs/asciidoc/faq.adoc [hunk 9]
```diff
@@ -85,3 +86,18 @@ Format
...
+or whether Reactor Netty already forwarded the ownership of the buffers to the application/framework.
+By default, this logger is disabled, in order to enable it - increase the log level to `DEBUG`.
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc:102]: Change to

By default, this logger is disabled. To enable it, increase the log level to `DEBUG`.

**Review by Buzzardo — CHANGES_REQUESTED** [review]

**Comment by violetagg** [comment]
@Buzzardo Thanks a lot! I applied the changes.

**Comment by violetagg** [comment]
@simonbasle Please check again the documentation

File: docs/asciidoc/faq.adoc [hunk 10]
```diff
@@ -85,3 +86,18 @@ Format
 Reactor Netty adds information for the connection at the beginning of every log record (when this is possible).
 Use the id of the connection in order to extract all log records for a particular HTTP request.
 For more information see <<faq.logging-prefix>>
+
+[[faq.memory-leaks]]
+== How can I debug a memory leak?
+By default, Reactor Netty uses direct memory as this is more performant
+when there are many native I/O operations (working with sockets), as this can remove the copying operations.
+As allocation and deallocation are expensive operations, the buffers (used in Reactor Netty by default) are pooled objects.
```
Comment (simonbasle) [inline docs/asciidoc/faq.adoc:94]: I would actually suggest to turn this sentence to:
```suggestion
As allocation and deallocation are expensive operations, Reactor Netty also uses pooled buffers by default.
```

**Review by simonbasle — CHANGES_REQUESTED** [review]

**Review by simonbasle — APPROVED** [review]

File: docs/asciidoc/faq.adoc [hunk 11]
```diff
@@ -85,3 +86,18 @@ Format
...
+As allocation and deallocation are expensive operations, Reactor Netty also uses pooled buffers by default.
+For more information, see https://github.com/netty/netty/wiki/Reference-counted-objects[Reference Counted Objects].
+
+To be able to debug memory issues with the direct memory and the pooled buffers, Netty provides a special memory leak detection mechanism.
+Follow the instructions for https://github.com/netty/netty/wiki/Reference-counted-objects#troubleshooting-buffer-leaks[Troubleshooting Buffer Leaks]
+to enable this mechanism. In addition to the instructions provided by Netty, Reactor Netty provides a special
+logger (`_reactor.netty.channel.LeakDetection`) that helps to identify where the memory leak might be located inside Reactor Netty
+or whether Reactor Netty already forwarded the ownership of the buffers to the application/framework.
+By default, this logger is disabled. To enable it, increase the log level to `DEBUG`.
+
```
Comment (nkonev) [inline docs/asciidoc/faq.adoc:103]: In addition I think it has a sense to mention reducing memory: lesser memory application has sooner memory leak will occur.

Some time ago I used
`-Djdk.nio.maxCachedBufferSize` and `-XX:MaxDirectMemorySize` for it when debug gateway.

Also there is an article https://dzone.com/articles/troubleshooting-problems-with-native-off-heap-memo

Comment (violetagg) [reply docs/asciidoc/faq.adoc:104]: @reactor/netty-team @nkonev I added a note for reducing the used memory. Please check.

**Review by simonbasle — APPROVED** [review]
LGTM, thanks for the memory config suggestion @nkonev

**Comment by violetagg** [comment]
> Maybe NMT ([Native Memory Tracking](https://docs.oracle.com/javase/8/docs/technotes/guides/troubleshoot/tooldescr007.html#:~:text=Start%20the%20JVM%20with%20summary,native_memory%20baseline%20.)) jvm options should be mentioned in the faq ? NMT provides jvm options and jcmd commands allowing to monitor internal jvm memory usage:
> 
> Shortly, to detect a native memory leak:
> 
> * start the jvm using -XX:+UnlockDiagnosticVMOptions -XX:NativeMemoryTracking=detail
> * start a load test, in order to warmup the jvm.
> * take a snapshot of current memory state using "jcmd  VM.native_memory baseline" command. It will be used as baseline.
> * redo another load test
> * and display memory changes using "jcmd  VM.native_memory detail.diff" command, which will display the difference in native memory usage since the baseline was taken.

Should we do this addition in a separate PR. @pderop Would you like to work on this?

**Comment by pderop** [comment]
ok, I will do it.

**Comment by violetagg** [comment]
> ok, I will do it (after you will have merged this one, to avoid conflicts).

yep that's great! Let me merge that one

Thanks you all for the review and the suggestions!
